File: src/slingshot/stone.clj
(ns slingshot.stone)

(defprotocol Stone (data [this]))

(defn create [data]
  (let [stone (proxy [RuntimeException slingshot.stone.Stone] []
                (data [] data))]
    (->> (.getStackTrace stone) (drop 6) into-array (.setStackTrace stone))
    stone))

File: java/src/slingshot/Stone.java
package slingshot;

public class Stone extends RuntimeException {

    final Object context;

    public Stone(String message, Throwable cause, StackTraceElement[] stackTrace,
                 Object context) {
        super (message, cause);
        setStackTrace(stackTrace);
        this.context = context;
    }

    public Object getContext () {
        return context;
    }
}
